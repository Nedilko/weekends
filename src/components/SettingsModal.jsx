import PropTypes from 'prop-types'
import Modal from './Modal'

SettingsModal.propTypes = {
  onApply: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
}

function SettingsModal({ onApply, onCancel }) {
  return (
    <Modal title="Settings" onApply={onApply} onCancel={onCancel}>
      <section className="font-thin text-xl text-slate-600 dark:text-gray-400">
        Lorem, ipsum dolor sit amet consectetur adipisicing elit. Amet ullam
        saepe, error quo in odio accusantium sed consequatur cupiditate, culpa
        minima ipsa suscipit assumenda, consequuntur possimus? Corporis ducimus
        a possimus.
      </section>
    </Modal>
  )
}

export default SettingsModal
