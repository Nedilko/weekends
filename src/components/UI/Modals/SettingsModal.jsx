import PropTypes from 'prop-types'
import { useContext, useState } from 'react'

import Modal from './Modal'
import SettingsModalRow from './SettingsModalRow'
import TextInput from '../Inputs/TextInput'
import Toggle from '../Toggles/Toggle'
import SettingsContext from '../../../store/Settings'

SettingsModal.propTypes = {
  onApply: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  settings: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired,
}

function SettingsModal({ title, settings, onApply, onCancel }) {
  const [greetingsText, setGeetingsText] = useState(settings.greetingsText)
  const [day, setDay] = useState(settings.day)
  const [hour, setHour] = useState(settings.hour)
  const [useSystemTheme, setuUseSystemTheme] = useState(settings.useSystemTheme)

  const ctx = useContext(SettingsContext)

  return (
    <Modal
      title={title}
      onApply={() =>
        onApply({
          greetingsText,
          day: Number(day),
          hour: Number(hour),
          useSystemTheme,
        })
      }
      onCancel={onCancel}
    >
      <section className="flex flex-col font-thin text-xl text-slate-600 dark:text-gray-400">
        <SettingsModalRow>
          <div className="flex">Greetings text</div>
          <TextInput
            placeholder={ctx.theme}
            value={greetingsText}
            onChange={setGeetingsText}
          />
        </SettingsModalRow>
        <SettingsModalRow>
          <div className="flex">Day</div>
          <TextInput placeholder="5" value={String(day)} onChange={setDay} />
        </SettingsModalRow>
        <SettingsModalRow>
          <div className="flex">Hour</div>
          <TextInput placeholder="18" value={String(hour)} onChange={setHour} />
        </SettingsModalRow>
        <SettingsModalRow>
          <div className="flex">Use system theme</div>
          <Toggle checked={useSystemTheme} onChange={setuUseSystemTheme} />
        </SettingsModalRow>
      </section>
    </Modal>
  )
}

export default SettingsModal
