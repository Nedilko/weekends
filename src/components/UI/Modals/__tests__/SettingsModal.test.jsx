import SettingsModal from '@UI/Modals/SettingsModal'
import { render, screen, userEvent } from '@utils/test-utils'

describe('SettingsModal', () => {
  const handleApply = vi.fn()
  const handleCancel = vi.fn()

  const settings = {
    useSystemTheme: false,
    theme: 'light',
    greetingsText: 'sample greeting',
    day: 5,
    hour: 18,
    isFirstLoad: false,
  }

  beforeAll(() => {
    window.HTMLElement.prototype.scrollIntoView = vi.fn()
  })

  afterEach(() => {
    window.HTMLElement.prototype.scrollIntoView.mockClear()
  })

  afterAll(() => {
    window.HTMLElement.prototype.scrollIntoView.mockRestore()
  })

  it('should be in the document', async () => {
    const modalRoot = document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
    render(
      <SettingsModal
        settings={settings}
        title="sample title"
        onApply={handleApply}
        onCancel={handleCancel}
      />
    )
    expect(screen.getByTestId('settings-modal')).toMatchSnapshot()
    document.body.removeChild(modalRoot)
  })

  it('should toggle switcher', async () => {
    const modalRoot = document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
    const user = userEvent.setup()
    render(
      <SettingsModal
        settings={settings}
        title="sample title"
        onApply={handleApply}
        onCancel={handleCancel}
      />
    )
    const toggle = screen.getByRole('checkbox')
    await user.click(toggle)
    expect(toggle).toBeChecked()
    document.body.removeChild(modalRoot)
  })

  it('should handle Apply with new settings', async () => {
    const modalRoot = document.createElement('div')
    modalRoot.setAttribute('id', 'modal-root')
    document.body.appendChild(modalRoot)
    const user = userEvent.setup()
    render(
      <SettingsModal
        settings={settings}
        title="sample title"
        onApply={handleApply}
        onCancel={handleCancel}
      />
    )

    const greetingsTextInput = screen.getByPlaceholderText('Have a beer')
    await user.clear(greetingsTextInput)
    await user.type(greetingsTextInput, 'new greeting')
    await user.click(screen.getByText('Friday'))
    await user.click(screen.getByText('Saturday'))
    await user.click(screen.getByText('18:00'))
    await user.click(screen.getByText('17:00'))
    await user.click(screen.getByRole('checkbox'))
    await user.click(screen.getByRole('button', { name: 'Apply' }))
    expect(handleApply).toHaveBeenCalledWith({
      greetingsText: 'new greeting',
      day: 6,
      hour: 17,
      useSystemTheme: true,
    })
    document.body.removeChild(modalRoot)
  })
})
