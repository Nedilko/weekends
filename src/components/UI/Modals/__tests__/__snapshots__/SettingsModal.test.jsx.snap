// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsModal should be in the document 1`] = `
<section
  class="flex flex-col border-y py-2 text-xl font-thin text-slate-600 dark:text-gray-400"
  data-testid="settings-modal"
>
  <div
    class="my-1 flex items-center justify-start gap-5"
  >
    <div
      class="flex w-1/2"
    >
      Greetings text
    </div>
    <input
      class="border-slate-100 focus:border-slate-300 mx-2 flex rounded-md border px-2 text-xl font-thin text-slate-600 focus:outline-none dark:border-gray-700 dark:bg-zinc-900 dark:text-gray-400 dark:focus:border-gray-500"
      data-testid="greetings-text-input"
      maxlength="20"
      placeholder="Have a beer"
      type="text"
      value="sample greeting"
    />
  </div>
  <div
    class="my-1 flex items-center justify-start gap-5"
  >
    <div
      class="flex flex-row items-center"
    >
      <div
        class="mr-2 flex items-center justify-center"
      >
        Day
      </div>
      <div
        class="z-10 block"
      >
        <button
          class="flex h-6 items-center justify-between rounded-md border border-slate-100 pl-2 text-base font-thin focus:border-slate-300 focus:outline-none dark:border-gray-700 dark:focus:border-gray-500"
          type="button"
        >
          Friday
          <svg
            class="mr-1 h-4 w-4 text-gray-500 "
            fill="none"
            role="img"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 9l-7 7-7-7"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex flex-row items-center"
    >
      <div
        class="mr-2 flex items-center justify-center"
      >
        Hour
      </div>
      <div
        class="z-10 block"
      >
        <button
          class="flex h-6 items-center justify-between rounded-md border border-slate-100 pl-2 text-base font-thin focus:border-slate-300 focus:outline-none dark:border-gray-700 dark:focus:border-gray-500"
          type="button"
        >
          18:00
          <svg
            class="mr-1 h-4 w-4 text-gray-500 "
            fill="none"
            role="img"
            stroke="currentColor"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 9l-7 7-7-7"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="my-1 flex items-center justify-start gap-5"
  >
    <div
      class="flex w-1/2"
    >
      Use system theme
    </div>
    <label
      class="flex items-center px-1 text-xl font-thin text-slate-600 dark:text-gray-400"
    >
      <span
        aria-checked="false"
        class="ml-4 flex h-6 w-12 flex-shrink-0 items-center justify-between rounded-full px-1 duration-300 ease-in-out bg-gray-200 dark:bg-slate-700"
        role="checkbox"
      >
        <span
          class="absolute flex h-5 w-5 rounded-full bg-white text-yellow-500 shadow-md duration-300 ease-in-out dark:bg-slate-900 "
        />
      </span>
    </label>
  </div>
</section>
`;
